lib/src/abstract_scope_manager.dart

```diff
@@ -0,0 +1,41 @@
+// Copyright 2016 Workiva Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:opentracing/opentracing.dart';
+
+/// The ScopeManager abstracts both the activation of [Span] instances via
+/// ScopeManager.activate(Span, boolean)} and access to an active
+/// [Span]/[Scope] via ScopeManager.active.
+abstract class ScopeManager {
+  /// Make a Span instance active.
+  ///
+  /// Span that should become the active Span
+  /// When [finishSpanOnClose] is true, the [span] will automatically be
+  /// finished when Scope.close() is called.
+  /// Returns a Scope instance to control the end of the active period for the
+  /// [Span]. It is a programming error to neglect calling Scope.close() on
+  /// the returned instance.
+  Scope activate(Span span, bool finishSpanOnClose);
+
+  /// Return the currently active [Scope] which can be used to access the
+  /// currently active [Scope].span.
+  /// If there is a [Scope] non-null scope, its wrapped [Span] becomes an
+  /// implicit parent(as [referenceChildOf]) of any newly-created
+  /// [Span] at [AbstractTracer].startSpan.
+  ///
+  /// Returns the [Scope], or null if none could be found.
+  Scope get active;
+
+  set active(Scope value);
```

Comment:
There shouldn't be a setter here; this value gets set in `activate`,  and there is no setter in the [Java implementation](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/ScopeManager.java).

Reply:
Please see https://github.com/Workiva/basictracer-dart/pull/59/files#diff-7c439d7fe63b0d14ff72fcbc3d094df5R32 for the reason to include this setter.

This allows us to auto-nest scopes:

```dart
Span operationASpan = new BasicSpan('OperationA')
scopeManager.activate(operationASpan, finishSpanOnClose);
/// start Operation A
/// make a bunch of spans representing sub-steps of Operation A
Span operationA1 = new BasicSpan('operationA1');
scopeManager.activate(operationA1, finishSpanOnClose);
/// make a bunch of spans representing sub-steps of Operation A1
/// end Operation A1
scopeManager.active.close()
/// end Operation A 
scopeManager.active.close()

```

The resulting span structure would be similar to :

operationA
     |- spans for sub-steps
     |- operationA1
               |-spans for sub-steps


Thoughts?

Reply:
That makes sense. Could we document that the setter should only be used by Scope/ScopeManager implementations, and discourage using it otherwise?

We would have to add ignore comments wherever used, but I think it'd be worth it to annotate it as `@protected` as well.